inc/class-saml.php

```diff
@@ -378,6 +378,50 @@ public function authenticate( $user, $username, $password ) {
 		return null;
 	}
 
+	/**
+	 * Return username from attributes given.
+	 *
+	 * @param $attributes
+	 * @return mixed
+	 * @throws \Exception
+	 */
+	public function getUsernameByAttributes( $attributes ) {
+		if ( isset( $attributes[ self::SAML_MAP_FIELDS['uid'] ] ) ) {
+			return $attributes[ self::SAML_MAP_FIELDS['uid'] ][0];
+		}
+		if ( isset( $attributes['friendlyAttributes']['uid'] ) ) {
+			return $attributes['friendlyAttributes']['uid'][0];
+		}
+		if ( isset( $attributes[ self::SAML_MAP_FIELDS['eduPersonPrincipalName'] ] ) ) {
+			return strstr(
+				$attributes[ self::SAML_MAP_FIELDS['eduPersonPrincipalName'] ][0],
+				'@',
+				true
+			);
+		}
+		return false;
```

Comment:
does this mean that a username could potentially/ever be set as `false` if each none of the three previous if statements were triggered?

Reply:
I realize we're checking in parseAttributeStatement to make sure that at least one of the three is set, but just curious about this

Reply:
From [here](https://github.com/pressbooks/pressbooks-saml-sso/pull/53/files#diff-cdef3d0de1baaf86b3e5d75d1a4e448ba10851e3425159bf36e18c91d09382b6R494) we know that case is not possible, at the moment to parse the attributes. I avoided an exception in case this function could be used in the future for other purposes.
Otherwise, the line could be deleted, or we could add an exception (I think it twice, because if we re use this function, probably won't want an exception,  that could be handled outside this action).